src/JustEat.StatsD/TagsFormatters/StatsDTagsFormatter.cs

```diff
@@ -27,53 +38,135 @@ protected StatsDTagsFormatter(StatsDTagsFormatterConfiguration configuration)
+            _prefixSize = Encoding.UTF8.GetByteCount(_prefix);
+            _suffixSize = Encoding.UTF8.GetByteCount(_suffix);
+            _tagsSeparatorSize = Encoding.UTF8.GetByteCount(_tagsSeparator);
+            _keyValueSeparatorSize = Encoding.UTF8.GetByteCount(_keyValueSeparator);
         }
         
         /// <inheritdoc />
         public bool AreTrailing { get; }
 
         /// <inheritdoc />
-        public virtual int GetTagsBufferSize(in IDictionary<string, string?> tags)
+        public virtual int GetTagsBufferSize(in Dictionary<string, string?> tags)
         {
             const int NoTagsSize = 0;
             if (!AreTagsPresent(tags))
             {
                 return NoTagsSize;
             }
 
-            return _prefix.Length
-                + Encoding.UTF8.GetByteCount(FormatTags(tags!))
-                + _suffix.Length;
+            return _prefixSize
+                + GetTagsSize(tags)
+                + _suffixSize;
         }
         
         /// <inheritdoc />
-        public virtual string FormatTags(in IDictionary<string, string?> tags)
+        public virtual ReadOnlySpan<char> FormatTags(in Dictionary<string, string?> tags)
         {
-            if (AreTagsPresent(tags))
+            if (!AreTagsPresent(tags))
+            {
+                return ReadOnlySpan<char>.Empty;
+            }
+
+            char[] destination = Buffer;
+            if (!TryFormatTags(destination, tags, out int written))
             {
-                return _prefix + FormatTags(tags!) + _suffix;
+                var newSize = GetTagsBufferSize(tags);
+                _buffer = new char[newSize];
+                destination = Buffer;
+
+                if (!TryFormatTags(destination, tags, out written))
+                {
+                    throw new Exception("Failed to format tags.");
+                }
             }
 
-            return string.Empty;
+            return new ArraySegment<char>(destination, 0, written);
+        }
+        
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private bool TryFormatTags(Span<char> destination, in Dictionary<string, string?> tags, out int written)
+        {
+            var buffer = new Buffer<char>(destination);
+            bool isFormattingSuccessful = buffer.TryWrite(_prefix)
+                && TryWriteTags(ref buffer, tags)
+                && buffer.TryWrite(_suffix);
+
+            written = isFormattingSuccessful ? buffer.Written : 0;
+
+            return isFormattingSuccessful;
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private string FormatTags(IDictionary<string, string?> tags) =>
-            string.Join(_tagsSeparator,tags.Select(FormatTags));
+        private bool TryWriteTags(ref Buffer<char> src, in Dictionary<string, string?> tags)
+        {
+            var enumerator = tags.GetEnumerator();
```

Comment:
Do you need to dispose of the enumerator?

Reply:
Done!